uuid/uuid.mbt

```diff
@@ -0,0 +1,185 @@
+// Copyright 2024 International Digital Economy Academy
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// UUID that complies with RFC 4122.
+/// 
+/// UUID objects are immutable, hashable and usable as mapping keys.
+/// The Display of UUIDs is like `12345678-1234-1234-1234-123456789abc`;
+/// UUIDs can be constructed using such strings, or 16-bytes strings with big-
+/// endian byte order.
+/// 
+/// # Usage
+/// 
+/// ```moonbit
+/// let u = @uuid.from_hex("ddf99703-742f-7505-4c54-df36a9c243fe")
+/// let u2 = @uuid.from_bytes(Bytes::[221, 249, 151, 3, 116, 47, 117, 5, 76, 84, 223, 54, 169, 194, 67, 254])
+/// @assertion.assert_eq(u, u2)
+/// ```
+/// 
+/// Construct an RFC-4122-compliant v4 UUID:
+/// 
+/// ```moonbit
+/// let u = @uuid
+///   .from_hex("9558cfadb48547c70f9adf1f75cdef29")
+///   .as_version(@uuid.UUID4)
+/// ```
+/// 
+struct UUID {
+  // XXX: replace with an 128-bit data structure when there is one
+  hi : Int64
+  lo : Int64
+}
+
+/// Constructs a UUID with the given big-endian bytes.
+/// 
+/// The produced UUID is not RFC-4122 compliant; see also `as_version()`.
+/// 
+pub fn from_bytes(bytes : Bytes) -> Result[UUID, String] {
+  if bytes.length() != 16 {
+    return Err("requires 16 bytes")
+  }
+  // XXX: this is ideally Int64::from_bytes() (needs a cheap slicing method)
+  let hi = for rv = 0L, i = 0; i < 8; {
+    continue rv.lor(bytes[i].to_int64().lsl((7 - i) * 8)), i + 1
+  } else {
+    rv
+  }
+  let lo = for rv = 0L, i = 8; i < 16; {
+    continue rv.lor(bytes[i].to_int64().lsl((15 - i) * 8)), i + 1
+  } else {
+    rv
+  }
+  Ok({ hi, lo })
+}
+
+/// The UUID as a 16-byte string.
+pub fn to_bytes(self : UUID) -> Bytes {
+  // XXX: replace with a cheaper way
+  let rv = Bytes::make(16, 0)
+  for i = 0; i < 8; i = i + 1 {
+    rv[i] = self.hi.lsr((7 - i) * 8).to_int()
+  }
+  for i = 8; i < 16; i = i + 1 {
+    rv[i] = self.lo.lsr((15 - i) * 8).to_int()
+  }
+  rv
+}
+
+test "bytes" {
+  let buf = Bytes::[
+    201, 130, 251, 104, 223, 80, 76, 161, 145, 248, 186, 162, 4, 9, 229, 185,
+  ]
+  inspect(from_bytes(buf)?.to_bytes(), ~content="苉棻僟ꅌꊺऄ맥")? // XXX: this is a terrible Show
```

Comment:
`mbt test -u` generated this crap - again, maybe using `Show` in `inspect()` is not the best option, we might need a new trait or something?

Reply:
cc @lijunchen is still the case?

Reply:
Yes. This is because the behavior of `Bytes::to_string` is to interpret the content as UTF-16 string.

Reply:
> Yes. This is because the behavior of `Bytes::to_string` is to interpret the content as UTF-16 string.

Maybe we need to change it to print bytes literal.